pkg/controller/certificaterequests/acme/acme.go

```diff
@@ -147,6 +147,15 @@ func (a *ACME) Sign(ctx context.Context, cr *v1alpha2.CertificateRequest, issuer
 
 		return nil, err
 	}
+	if !metav1.IsControlledBy(order, cr) {
```

Comment:
So if I understand correctly we can get in a state where multiple CRs are waiting for the same order and therefore propagate up a cert only valid for one in the set. 

This PR fixes that by queuing up the CRs and has the queue in order of who wins the race. 

What I don't understand is how the order is ever deleted after the first CR success with this line here.

Reply:
> So if I understand correctly we can get in a state where multiple CRs are waiting for the same order and therefore propagate up a cert only valid for one in the set.

Not quite, the exact case is a bit of a race:

* Assume a Certificate exists, is issued and is valid - it owns a corresponding 'CertificateRequest' which is in a 'Ready' state, which in turn owns an 'Order' resource which is in a 'valid' state.
* If a user deletes the *Secret* resource associated with the Certificate, the private key (and certificate) is therefore lost, and the Certificate controller goes to trigger a re-issuance
* It generates a new private key and stores it in the Secret resource (at this point, *only* tls.key is set as the user had deleted the old Secret resource)
* It checks the existing CertificateRequest and determines that the CSR is *not* signed by the public key that has just been generated (as it is left-over/old) (here: https://github.com/jetstack/cert-manager/blob/6d9200f9d32aa158fe600ff4d0883763840b1e71/pkg/controller/certificates/sync.go#L392-L396) - this is correct behaviour
* The Certificates controller then creates a new CertificateRequest resource with the same name as the one it previously deleted, as the CR name hash function does not account for the public key when generating the hash
* (this is where the race happens) the ACME CertificateRequest controller then attempts to Get an Order with the name computed by the `hashOrder` function, which *also* does not account for the public key in the hash function
* It finds the *old* Order resource, that is owned by the *old* CertificateRequest resource, because the names match
* (this is a bug that this PR fixes) - the ACME CR sees that this Order is 'valid' and has a `status.certificate` set, and (as you can see here: https://github.com/jetstack/cert-manager/blob/6d9200f9d32aa158fe600ff4d0883763840b1e71/pkg/controller/certificaterequests/acme/acme.go#L164-L171) automatically copies `status.certificate` from the Order to `status.certificate` on the CertificateRequest - it *does not actually validate that the Certificate's private key is valid for the CSR being requested*
* At this point, we now have a CertificateRequest resource that is marked as Ready, with a `spec.csr` that is signed using the *new* private key, but the `status.certificate` signed by the *old private key* as it was copied from an *old* Order resource
* (this is a second bug fixed by this PR), the Certificates controller *should* be able to handle this situation, albeit the situation shouldn't really occur in the first place. *HOWEVER*, it currently does *not* validate that `status.certificate` is valid for the private key we have stored when copying from the CertificateRequest: https://github.com/jetstack/cert-manager/blob/6d9200f9d32aa158fe600ff4d0883763840b1e71/pkg/controller/certificates/sync.go#L459-L472

The infinite issuance comes about because the Certificates controller *does* check whether the public/private key is up to date in a Secret here: https://github.com/jetstack/cert-manager/blob/6d9200f9d32aa158fe600ff4d0883763840b1e71/pkg/controller/certificates/sync.go#L269-L270 - this area of code correctly says the public keys do not match, meaning we go into a loop of detecting something doesn't match but then re-issuing that same certificate.

This issue only comes up in the very specific case where this is a race condition in a lower down controller (i.e. the CertificateRequest controller) that means that an invalid 'pair' exists between `spec.csr` and `status.certificate` - the Certificate's controller *does* validate that the *CSR's* public key is valid for the private key (here: https://github.com/jetstack/cert-manager/blob/6d9200f9d32aa158fe600ff4d0883763840b1e71/pkg/controller/certificates/sync.go#L384-L396), but as I noted, the bug in the *CertificateRequest* controller means that in certain cases these two public keys can be out of sync, hence surfacing both bugs.

Reply:
> What I don't understand is how the order is ever deleted after the first CR success with this line here.

The reason why this works, is that the garbage collection controller will, a few hundred ms later, delete the problematic Order resource.

I opted to return an error here instead of deleting it directly (like I've done a few lines below this one) because in this particular instance, the Order is not *owned* by this CertificateRequest so it seems improper for us to delete it and potentially 'step on toes'.

This weirdness is what led me to suggest we should move *away* from using hash based resource names, as it means we can run into conflict cases where a resource already exists with that name but it is not owned by 'us', meaning we cannot proceed.

Reply:
Ah that's perfect, thanks a lot for writing all that up